### PR TITLE
Another round of v1alpha2 cleanup

### PR DESCRIPTION
**Pull request by robscott**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This takes care of another big chunk of feedback from #780 as well as some of the items in #790.

**Does this PR introduce a user-facing change?**:
```release-note
* "Controller" has been renamed to "ControllerName"
* "Admitted" condition has been renamed to "Accepted" and now defaults to an "Unknown" state instead of "False"
```

### CODE REVIEW

File: apis/v1alpha2/tlsroute_types.go [hunk 1]
```diff
@@ -49,26 +49,35 @@ type TLSRouteSpec struct {
...
-	// 2. If SNI is a wildcard, then the request matches this Route if the
-	//    SNI is to equal to the suffix (removing the first label) of the
-	//    wildcard.
-	// 3. If SNIs are unspecified, all requests associated with the gateway TLS
-	//    listener will match. This can be used to define a default backend
-	//    for a TLS listener.
+	// SNI attribute of TLS ClientHello message in TLS handshake. This matches
+	// the RFC 1123 definition of a hostname with 2 notable exceptions:
+	//
+	// 1. IPs are not allowed.
```
Comment (howardjohn) [inline apis/v1alpha2/tlsroute_types.go:55]: IPs are not allowed in SNI in the RFC

File: apis/v1alpha2/tlsroute_types.go [hunk 2]
```diff
@@ -49,26 +49,35 @@ type TLSRouteSpec struct {
...
+	//
+	// If both the Listener and TLSRoute have specified hostnames, any
+	// TLSRoute hostnames that do not match the Listener hostname MUST be
+	// ignored. For example, if a Listener specified `*.example.com`, and the
+	// TLSRoute specified `test.example.com` and `test.example.net`,
+	// `test.example.net` must not be considered for a match.
+	//
+	// If hostnames do not match with the criteria above, then the TLSRoute is
+	// not accepted, and the implementation must raise an 'Accepted' Condition
+	// with a status of `False` for the target Listener(s).
```
Comment (howardjohn) [inline apis/v1alpha2/tlsroute_types.go:80]: This is a bit confusing when not using `sectionName` in parentRefs. Since the target is not a listener

File: apis/v1alpha2/tlsroute_types.go [hunk 3]
```diff
@@ -49,26 +49,35 @@ type TLSRouteSpec struct {
...
+	//   `test.example.com` and `*.example.com` would both match. On the other
+	//   hand, `example.com` and `test.example.net` would not match.
+	//
+	// If both the Listener and TLSRoute have specified hostnames, any
+	// TLSRoute hostnames that do not match the Listener hostname MUST be
+	// ignored. For example, if a Listener specified `*.example.com`, and the
+	// TLSRoute specified `test.example.com` and `test.example.net`,
+	// `test.example.net` must not be considered for a match.
+	//
+	// If hostnames do not match with the criteria above, then the TLSRoute is
```
Comment (howardjohn) [inline apis/v1alpha2/tlsroute_types.go:78]: I would clarify: `If <any,all> hostnames ...`. And should be All IMO

File: apis/v1alpha2/gateway_types.go [hunk 4]
```diff
@@ -72,6 +72,9 @@ type GatewaySpec struct {
 	// logical endpoints that are bound on this Gateway's addresses.
 	// At least one Listener MUST be specified.
 	//
+	// Each listener in a Gateway must have a unique combination of Hostname,
+	// Port, and Protocol. This will be enforced by a validating webhook.
```
[the remark this answers is not in the document]
Comment (hbagdi) [reply apis/v1alpha2/gateway_types.go:76]: Recommend adding an issue instead.

File: apis/v1alpha2/gateway_types.go [hunk 5]
```diff
@@ -143,9 +146,7 @@ type GatewaySpec struct {
 }
 
 // Listener embodies the concept of a logical endpoint where a Gateway can
-// accept network connections. Each listener in a Gateway must have a unique
-// combination of Hostname, Port, and Protocol. This will be enforced by a
-// validating webhook.
+// accept network connections.
```
Comment (hbagdi) [inline apis/v1alpha2/gateway_types.go:149]: // Listener embodies the concept of a logical endpoint where a Gateway ~can~ accepts network connections.

File: apis/v1alpha2/gateway_types.go [hunk 6]
```diff
@@ -157,10 +158,22 @@ type Listener struct {
...
+	// the following protocols:
+	//
+	// * TLS: The Hostname match MUST be applied to the
+	//   [SNI](https://tools.ietf.org/html/rfc6066#section-3) server name
+	//   offered by the client.
+	// * HTTP: The Hostname match MUST be applied to the host portion of the
+	//   [effective request URI](https://tools.ietf.org/html/rfc7230#section-5.5)
+	//   or the [:authority pseudo-header](https://tools.ietf.org/html/rfc7540#section-8.1.2.3)
+	// * HTTPS: The Hostname match MUST be applied at both the TLS and HTTP
+	//   protocol layers.
```
Comment (hbagdi) [inline apis/v1alpha2/gateway_types.go:171]: @youngnick Is this sufficient to close the domain fronting issue (#72) as well?

File: apis/v1alpha2/tcproute_types.go [hunk 7]
```diff
@@ -27,7 +27,9 @@ import (
 // +kubebuilder:storageversion
 // +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
 
-// TCPRoute is the Schema for the TCPRoute resource.
+// TCPRoute provides a way to route TCP requests. When combined with a Gateway
+// listener, it can be used to forward requests to the port specified by the
```
Comment (hbagdi) [inline apis/v1alpha2/tcproute_types.go:31]:
```suggestion
// listener, it can be used to forward connections on the port specified by the
```

File: apis/v1alpha2/tlsroute_types.go (hunk 1 above)
Comment (hbagdi) [reply apis/v1alpha2/tlsroute_types.go:55]: Any relevant use cases for supporting IPs?
Can we start narrow and open up later?

File: apis/v1alpha2/gateway_types.go (hunk 6 above)
Comment (robscott) [reply apis/v1alpha2/gateway_types.go:171]: Although this isn't entirely new content, now that I'm seeing it here, I think this wording may be a bit too strong. I'm not sure that every implementation _can_ apply a match at both layers.

File: apis/v1alpha2/tlsroute_types.go (hunk 1 above)
Comment (robscott) [reply apis/v1alpha2/tlsroute_types.go:55]: I think you're both agreeing that IPs should not be allowed, but John is also saying we probably don't even need to specify that here.

File: apis/v1alpha2/gateway_types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (youngnick) [reply apis/v1alpha2/gateway_types.go:76]: I'd include the TLS details in that list as well, tbh.

File: apis/v1alpha2/gateway_types.go (hunk 6 above)
Comment (youngnick) [reply apis/v1alpha2/gateway_types.go:171]: I think the direction is good, but maybe we need to specify this in terms of API objects, like:
```
- TLS: the Listener hostname must match the SNI.
- HTTP: The HTTPRoute hostname must match the Host header of the request.
- HTTPS: The Listener Hostname and the HTTPRoute hostname must agree with each other, and also match request properties as above.
```

Comment (youngnick) [reply apis/v1alpha2/gateway_types.go:171]: Because the important part is that we expect that Listeners and HTTPRoutes that do not match these rules should not end up attached.

File: apis/v1alpha2/gateway_types.go [hunk 8]
```diff
@@ -610,17 +608,14 @@ type ListenerStatus struct {
...
-	// "InvalidRouteKinds" reason. If both valid and invalid Route kinds are
-	// specified, the implementation should support the valid Route kinds that
-	// have been specified.
+	// listener. This MUST represent the kinds an implementation supports for
+	// that Listener configuration.
+	//
+	// If kinds are specified in Spec that are not supported, they MUST NOT
+	// appear in this list and an implementation MUST set the "ResolvedRefs"
+	// condition to "False" with the "InvalidRouteKinds" reason. If both valid
+	// and invalid Route kinds are specified, the implementation should support
```
Comment (youngnick) [inline apis/v1alpha2/gateway_types.go:617]:
```suggestion
	// and invalid Route kinds are specified, the implementation should reference
```

File: apis/v1alpha2/shared_types.go [hunk 9]
```diff
@@ -230,19 +230,24 @@ type RouteStatus struct {
...
 	// means the route has not been attached to any Gateway.
 	//
 	// +kubebuilder:validation:MaxItems=32
 	Parents []RouteParentStatus `json:"parents"`
 }
 
-// Hostname is the fully qualified domain name of a network host, as defined
-// by RFC 3986. Note the following deviations from the "host" part of the
-// URI as defined in the RFC:
+// Hostname is the fully qualified domain name of a network hostThis matches
```
Comment (youngnick) [inline apis/v1alpha2/shared_types.go:245]:
```suggestion
// Hostname is the fully qualified domain name of a network host. This matches
```

File: apis/v1alpha2/tlsroute_types.go (hunk 1 above)
Comment (youngnick) [reply apis/v1alpha2/tlsroute_types.go:55]: This is copied from other places which abide exactly by RFC1123, which *does* say that it SHOULD be possible to enter an IP in dotted-decimal format as a hostname, so it's important in those cases.

I agree that since SNI explicitly disallows literal addresses in the hostname field, it's not actually needed here, but I think it doesn't hurt and makes the rules across the different types of resources in our API more similar.

In short, I think we should probably keep this.

File: apis/v1alpha2/tlsroute_types.go (hunk 3 above)
Comment (youngnick) [reply apis/v1alpha2/tlsroute_types.go:78]: +1 that it the Route should only not be accepted if *all* hostnames do not match, and if at least one does, a ResolvedRefs condition should be raised as well.

File: apis/v1alpha2/tlsroute_types.go (hunk 2 above)
Comment (youngnick) [reply apis/v1alpha2/tlsroute_types.go:80]: Yeah, this comes down to the behavior with partial Listener acceptance, I think if we mandate "all", and have a ResolvedRefs in the event not all match, then this covers the sectionName case.

**Review by youngnick — COMMENTED** [review]
Overall LGTM, some small comments.

File: apis/v1alpha2/gateway_types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (robscott) [reply apis/v1alpha2/gateway_types.go:76]: Created https://github.com/kubernetes-sigs/gateway-api/issues/847 to track webhook validation.

> I'd include the TLS details in that list as well, tbh.

Do you mean that hostname, listener, and protocol can be identical across listener if TLS is different?

File: apis/v1alpha2/tlsroute_types.go (hunk 2 above)
Comment (robscott) [reply apis/v1alpha2/tlsroute_types.go:80]: I agree that we need a better way to handle status when `sectionName` is not specified. 2 options I've been thinking of:

1. A standard/reserved listener name such as "all" that is used in status to cover failed attachment to the Gateway as a whole.
2. A new set of condition types for Gateway status specifically for this purpose. (We could potentially reuse the existing listener condition types).

File: apis/v1alpha2/gateway_types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (robscott) [reply apis/v1alpha2/gateway_types.go]: Removed reference to webhook in godocs here.

File: apis/v1alpha2/gateway_types.go (hunk 6 above)
Comment (robscott) [reply apis/v1alpha2/gateway_types.go]: I don't think we should mix up the route type with the listener protocol here. While there's likely a ~99% overlap here, if we tie route types too closely here, we prevent or at least complicate extended route types. I've copied most of your suggestions from above, but removed any references to Routes - the following paragraph already covers that intersection.

File: apis/v1alpha2/gateway_types.go (hunk 8 above)
Comment (robscott) [reply apis/v1alpha2/gateway_types.go]: Updated, thanks!

File: apis/v1alpha2/gatewayclass_types.go [hunk 10]
```diff
@@ -130,8 +126,8 @@ type ParametersReference struct {
 	Scope *string `json:"scope,omitempty"`
 
 	// Namespace is the namespace of the referent.
-	// This field is required when scope is set to "Namespace" and ignored when
-	// scope is set to "Cluster".
+	// This field is required when scope is set to "Namespace" must be unset
```
[the remark this answers is not in the document]
Comment (robscott) [reply apis/v1alpha2/gatewayclass_types.go]: Updated, thanks!

File: apis/v1alpha2/shared_types.go (hunk 9 above)
Comment (robscott) [reply apis/v1alpha2/shared_types.go]: Updated, thanks!

File: apis/v1alpha2/tcproute_types.go (hunk 7 above)
Comment (robscott) [reply apis/v1alpha2/tcproute_types.go]: Updated, thanks!

File: apis/v1alpha2/tlsroute_types.go (hunk 1 above)
Comment (robscott) [reply apis/v1alpha2/tlsroute_types.go]: Kept in place since the starting point is RFC 1123, but added a reference to RFC 6066 for this bullet point to explain that IP addresses were not allowed in SNI names.

File: apis/v1alpha2/tlsroute_types.go (hunk 3 above)
Comment (robscott) [reply apis/v1alpha2/tlsroute_types.go]: >+1 that it the Route should only not be accepted if all hostnames do not match

Updated, thanks!

> and if at least one does, a ResolvedRefs condition should be raised as well.

Wouldn't this result in a successful attachment? Why would we raise a ResolvedRefs condition?

**Comment by robscott** [comment]
@youngnick @hbagdi @howardjohn Thanks for the great feedback on this! I think I've responded to everything, PTAL.

File: apis/v1alpha2/gateway_types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (youngnick) [reply apis/v1alpha2/gateway_types.go]: > Do you mean that hostname, listener, and protocol can be identical across listener if TLS is different?

Oh, you're quite right, that makes no sense. I was more thinking that TLS details are part of the key for this, but you're right that if the hostname, port, and protocol are identical, then the TLS details must not be different.

File: apis/v1alpha2/tlsroute_types.go (hunk 2 above)
Comment (youngnick) [reply apis/v1alpha2/tlsroute_types.go]: I think option 2 seems better - if no listeners are valid, that seems very relevant for top-level status, but I would have thought we had that covered already somehow.

File: apis/v1alpha2/tlsroute_types.go (hunk 3 above)
Comment (youngnick) [reply apis/v1alpha2/tlsroute_types.go]: I guess you're right, the thing I wanted to make sure we have is to indicate what hostnames *are* in use somehow. If you've got a bunch of hostnames and only one is in use, that seems like it would be good to know.

File: apis/v1alpha2/httproute_types.go [hunk 11]
```diff
@@ -487,6 +484,10 @@ type HTTPRouteFilter struct {
 	// Implementers are encouraged to define custom implementation types to
 	// extend the core API with implementation-specific behavior.
 	//
+	// If a reference to a custom filter type cannot be resolved, the filter
+	// MUST NOT be skipped. Instead, requests that would have been processed by
+	// that filter SHOULD receive a HTTP error response.
```
Comment (youngnick) [inline apis/v1alpha2/httproute_types.go:489]: Why would they not do this? If it's only because the implementation doesn't support it, then we should say "MUST, except if".

**Review by youngnick — COMMENTED** [review]
LGTM aside from me being nitpicky about SHOULD.

I think that we need to ensure that every time we use SHOULD, we clearly explain the reasons that is not MUST, and consider using "MUST, except if" instead.

I think that SHOULD leaves too much room for interpretation without clarification.

File: apis/v1alpha2/gateway_types.go [hunk 12]
```diff
@@ -157,10 +158,18 @@ type Listener struct {
 	// field is ignored for protocols that don't require hostname based
 	// matching.
 	//
+	// Implementations MUST apply Hostname matching appropriately for each of
+	// the following protocols:
+	//
+	// * TLS: The Listener Hostname MUST match the SNI.
+	// * HTTP: The Listener Hostname MUST match the Host header of the request.
+	// * HTTPS: The Listener Hostname SHOULD match at both the TLS and HTTP
+	//   protocol layers as described above.
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply apis/v1alpha2/gateway_types.go:167]: At the very least, the SNI MUST match the listener hostname I would think. And the HTTP Host MUST match the HTTPRoute hostname. And the listener/route hostnames MUST form an intersection. Therefor, there MUST be some intersection between the two if both are specified. For example, if I have listener hostname `*.example.com`, then both MUST have `*.example.com`.

Not that we should write that as is.. just brain dumping.

File: apis/v1alpha2/tlsroute_types.go (hunk 3 above)
Comment (robscott) [reply apis/v1alpha2/tlsroute_types.go]: https://github.com/kubernetes-sigs/gateway-api/issues/848

File: apis/v1alpha2/tlsroute_types.go (hunk 2 above)
Comment (robscott) [reply apis/v1alpha2/tlsroute_types.go]: https://github.com/kubernetes-sigs/gateway-api/issues/849

File: apis/v1alpha2/httproute_types.go (hunk 11 above)
Comment (robscott) [reply apis/v1alpha2/httproute_types.go]: Yeah I think you're right, can't think of a reason not to use MUST here.

File: apis/v1alpha2/gateway_types.go [hunk 13]
```diff
@@ -611,17 +605,14 @@ type ListenerStatus struct {
...
-	// specified, the implementation should support the valid Route kinds that
-	// have been specified.
+	// listener. This MUST represent the kinds an implementation supports for
+	// that Listener configuration.
+	//
+	// If kinds are specified in Spec that are not supported, they MUST NOT
+	// appear in this list and an implementation MUST set the "ResolvedRefs"
+	// condition to "False" with the "InvalidRouteKinds" reason. If both valid
+	// and invalid Route kinds are specified, the implementation should
+	// reference the valid Route kinds that have been specified.
```
[the remark this answers is not in the document]
Comment (robscott) [reply apis/v1alpha2/gateway_types.go]: Updated, thanks!

File: apis/v1alpha2/gateway_types.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (robscott) [reply apis/v1alpha2/gateway_types.go]: I took a shot at updating this. What I'm trying to communicate here is that if implementations can implement the match at both levels they should, but not every implementation can.

**Comment by youngnick** [comment]
I think you got it @robscott, nice work.

/lgtm

/hold
for another lgtm though.

File: apis/v1alpha2/gateway_types.go [hunk 14]
```diff
@@ -718,8 +711,11 @@ const (
 	// interoperability.
 	ListenerConditionDetached ListenerConditionType = "Detached"
 
-	// This reason is used with the "Detached" condition when the
-	// Listener requests a port that cannot be used on the Gateway.
+	// This reason is used with the "Detached" condition when the Listener
+	// requests a port that cannot be used on the Gateway. This reason could be
+	// used in a number of instances, including:
```
Comment (jpeach) [inline apis/v1alpha2/gateway_types.go:716]: Need a newline here to render the markdown bulleted list.

File: apis/v1alpha2/gateway_types.go [hunk 15]
```diff
@@ -733,9 +729,11 @@ const (
 	// protocol type is not supported.
 	ListenerReasonUnsupportedProtocol ListenerConditionReason = "UnsupportedProtocol"
 
-	// This reason is used with the "Detached" condition when
-	// the Listener could not be attached to the Gateway because the
-	// requested address is not supported.
+	// This reason is used with the "Detached" condition when the Listener could
+	// not be attached to the Gateway because the requested address is not
+	// supported. This reason could be used in a number of instances, including:
```
Comment (jpeach) [inline apis/v1alpha2/gateway_types.go:734]: Need a newline here to render the markdown bulleted list.

File: apis/v1alpha2/httproute_types.go [hunk 16]
```diff
@@ -455,16 +453,15 @@ type HTTPRouteMatch struct {
...
-// +union
+// HTTPRouteFilter defines processing steps that must be completed during the
+// request or response lifecycle. HTTPRouteFilters are meant as an extension
+// point to express processing that may be done in Gateway implementations. Some
+// examples include request or response modification, implementing
+// authentication strategies, rate-limiting, and traffic shaping. API
+// guarantee/conformance is defined based on the type of the filter.
+// TODO(hbagdi): re-render CRDs once controller-tools supports union tags: -
+// https://github.com/kubernetes-sigs/controller-tools/pull/298 -
+// https://github.com/kubernetes-sigs/controller-tools/issues/461 +union
```
Comment (jpeach) [inline apis/v1alpha2/httproute_types.go:464]: The `+union` should be on a line by itself. Separate TODO into a separate paragraph, or drop it into an issue.

**Comment by jpeach** [comment]
Just a couple of formatting nits.

/lgtm

File: apis/v1alpha2/httproute_types.go (hunk 16 above)
Comment (robscott) [reply apis/v1alpha2/httproute_types.go]: Good call, removed entirely: https://github.com/kubernetes-sigs/gateway-api/issues/854.

File: apis/v1alpha2/gateway_types.go (hunk 15 above)
Comment (robscott) [reply apis/v1alpha2/gateway_types.go:738]: Updated, thanks!

File: apis/v1alpha2/gateway_types.go (hunk 14 above)
Comment (robscott) [reply apis/v1alpha2/gateway_types.go:719]: Updated, thanks!
